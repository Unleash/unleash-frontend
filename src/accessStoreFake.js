import { Map as $MAp } from 'immutable';

export const createFakeStore = (permissions) => {
    return {
        getState: () => ({
            user: 
                new $MAp({
                    permissions
                })
        }),
    }
}