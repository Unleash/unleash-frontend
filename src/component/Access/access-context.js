import React from 'react';

const AccessContext = React.createContext()

const AccessProvider = ({store, children}) => {
  const hasAccess = (permission, project) => {
    const permissions = store.getState().user.get('permissions');

    return permissions.some(p => {
      if(p.permission === 'ADMIN') {
        return true
      }
      if(p.permission === permission && p.project === project) {
        return true;
      }
      return false;
    })
    
  };

  const context = { hasAccess };

  return <AccessContext.Provider value={context}>{children}</AccessContext.Provider>
}

export {Â AccessProvider };
export default AccessContext;
