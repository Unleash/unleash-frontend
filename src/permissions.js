export const ADMIN = 'ADMIN';
export const CREATE_FEATURE = 'CREATE_FEATURE';
export const UPDATE_FEATURE = 'UPDATE_FEATURE';
export const DELETE_FEATURE = 'DELETE_FEATURE';
export const CREATE_STRATEGY = 'CREATE_STRATEGY';
export const UPDATE_STRATEGY = 'UPDATE_STRATEGY';
export const DELETE_STRATEGY = 'DELETE_STRATEGY';
export const UPDATE_APPLICATION = 'UPDATE_APPLICATION';
export const CREATE_CONTEXT_FIELD = 'CREATE_CONTEXT_FIELD';
export const UPDATE_CONTEXT_FIELD = 'UPDATE_CONTEXT_FIELD';
export const DELETE_CONTEXT_FIELD = 'DELETE_CONTEXT_FIELD';

export function hasPermission(user, permission) {
    return (
        user &&
        (!user.permissions || user.permissions.indexOf(ADMIN) !== -1 || user.permissions.indexOf(permission) !== -1)
    );
}
