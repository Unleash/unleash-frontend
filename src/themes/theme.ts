import { createTheme } from '@mui/material/styles';
import { colors } from './colors';

export default createTheme({
    boxShadows: {
        main: '0px 2px 4px rgba(129, 122, 254, 0.2)',
    },
    typography: {
        fontFamily: 'Sen, Roboto, sans-serif',
        fontWeightBold: '700',
        fontWeightMedium: '700',
        allVariants: { lineHeight: 1.4 },
        button: { lineHeight: 1.75 },
        h1: {
            fontSize: '1.5rem',
            lineHeight: 1.875,
        },
    },
    fontSizes: {
        mainHeader: '1.25rem',
        subHeader: '1.1rem',
        bodySize: '1rem',
        smallBody: '0.9rem',
        smallerBody: '0.8rem',
    },
    fontWeight: {
        thin: 300,
        medium: 400,
        semi: 700,
        bold: 700,
    },
    shape: {
        borderRadius: '3px',
        borderRadiusLarge: '10px',
        borderRadiusExtraLarge: '25px',
    },
    palette: {
        primary: {
            main: colors.purple[800],
            light: colors.purple[700],
            dark: colors.purple[900],
        },
        secondary: {
            main: colors.purple[800],
            light: colors.purple[700],
            dark: colors.purple[900],
        },
        info: {
            light: colors.blue[700],
            main: colors.blue[700],
            dark: colors.blue[800],
        },
        success: {
            light: colors.green[700],
            main: colors.green[700],
            dark: colors.green[800],
        },
        warning: {
            light: colors.orange[700],
            main: colors.orange[700],
            dark: colors.orange[800],
        },
        error: {
            light: colors.red[700],
            main: colors.red[700],
            dark: colors.red[800],
        },
        divider: colors.grey[300],
        dividerAlternative: colors.grey[500],
        grey: colors.grey,
        text: {
            primary: colors.grey[900],
            secondary: colors.grey[800],
            disabled: colors.grey[600],
        },
        code: {
            main: '#0b8c8f',
            diffAdd: 'green',
            diffSub: 'red',
            diffNeutral: 'black',
            edited: 'blue',
            background: '#efefef',
        },
        activityIndicators: {
            unknown: colors.grey[100],
            recent: colors.green[100],
            inactive: colors.orange[200],
            abandoned: colors.red[200],
        },
    },
    components: {
        MuiTableHead: {
            styleOverrides: {
                root: {
                    background: colors.grey[200],
                },
            },
        },
        MuiLink: {
            styleOverrides: {
                root: {
                    color: colors.purple[900],
                },
            },
        },
        MuiBreadcrumbs: {
            styleOverrides: {
                root: {
                    '& a': {
                        color: colors.purple[900],
                    },
                },
            },
        },
        MuiTableRow: {
            styleOverrides: {
                root: {
                    '&.MuiTableRow-hover:hover': {
                        background: colors.grey[100],
                    },
                },
            },
        },
    },
});
