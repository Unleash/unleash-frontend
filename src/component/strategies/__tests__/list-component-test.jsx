import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { ThemeProvider } from '@material-ui/core';

import StrategiesListComponent from '../StrategiesList/StrategiesList';
import renderer from 'react-test-renderer';
import theme from '../../../themes/main-theme';
import { AccessProvider } from '../../Access/access-context';
import { createFakeStore } from '../../../accessStoreFake';
import { ADMIN } from '../../Access/permissions';

test('renders correctly with one strategy', () => {
    const strategy = {
        name: 'Another',
        description: "another's description",
    };
    const tree = renderer.create(
        <MemoryRouter>
            <ThemeProvider theme={theme}>
                <AccessProvider store={createFakeStore()}>
                    <StrategiesListComponent
                        strategies={[strategy]}
                        fetchStrategies={jest.fn()}
                        removeStrategy={jest.fn()}
                        deprecateStrategy={jest.fn()}
                        reactivateStrategy={jest.fn()}
                        history={{}}
                    />
                </AccessProvider>
            </ThemeProvider>
        </MemoryRouter>
    );

    expect(tree).toMatchSnapshot();
});

test('renders correctly with one strategy without permissions', () => {
    const strategy = {
        name: 'Another',
        description: "another's description",
    };
    const tree = renderer.create(
        <MemoryRouter>
            <ThemeProvider theme={theme}>
                <AccessProvider store={createFakeStore([{permission: ADMIN}])}>
                <StrategiesListComponent
                    strategies={[strategy]}
                    fetchStrategies={jest.fn()}
                    removeStrategy={jest.fn()}
                    deprecateStrategy={jest.fn()}
                    reactivateStrategy={jest.fn()}
                    history={{}}
                />
                </AccessProvider>
            </ThemeProvider>
        </MemoryRouter>
    );

    expect(tree).toMatchSnapshot();
});
