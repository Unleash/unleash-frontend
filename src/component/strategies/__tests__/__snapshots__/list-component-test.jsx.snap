// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with one strategy 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
>
  <div
    className="makeStyles-headerContainer-3"
  >
    <div
      className="makeStyles-headerTitleContainer-7"
    >
      <div
        className=""
      >
        <h2
          className="MuiTypography-root makeStyles-headerTitle-8 MuiTypography-h2"
        >
          Strategies
        </h2>
      </div>
      <div
        className="makeStyles-headerActions-9"
      />
    </div>
  </div>
  <div
    className="makeStyles-bodyContainer-4"
  >
    <ul
      className="MuiList-root MuiList-padding"
    >
      <li
        className="MuiListItem-root makeStyles-listItem-1 MuiListItem-gutters"
        disabled={false}
      >
        <div
          className="MuiListItemAvatar-root"
        >
          <span
            aria-hidden={true}
            className="material-icons MuiIcon-root"
          >
            extension
          </span>
        </div>
        <div
          className="MuiListItemText-root MuiListItemText-multiline"
        >
          <span
            className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            <a
              href="/strategies/view/Another"
              onClick={[Function]}
            >
              <strong>
                Another
              </strong>
            </a>
          </span>
          <p
            className="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
          >
            another's description
          </p>
        </div>
        <div
          aria-describedby={null}
          className=""
          onBlur={[Function]}
          onFocus={[Function]}
          onMouseLeave={[Function]}
          onMouseOver={[Function]}
          onTouchEnd={[Function]}
          onTouchStart={[Function]}
          title="Deprecate activation strategy"
        >
          <button
            className="MuiButtonBase-root MuiIconButton-root"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiIconButton-label"
            >
              <span
                aria-hidden={true}
                className="material-icons MuiIcon-root"
              >
                visibility_off
              </span>
            </span>
            <span
              className="MuiTouchRipple-root"
            />
          </button>
        </div>
      </li>
    </ul>
  </div>
</div>
`;

exports[`renders correctly with one strategy without permissions 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
>
  <div
    className="makeStyles-headerContainer-3"
  >
    <div
      className="makeStyles-headerTitleContainer-7"
    >
      <div
        className=""
      >
        <h2
          className="MuiTypography-root makeStyles-headerTitle-8 MuiTypography-h2"
        >
          Strategies
        </h2>
      </div>
      <div
        className="makeStyles-headerActions-9"
      >
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            Add new strategy
          </span>
          <span
            className="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    className="makeStyles-bodyContainer-4"
  >
    <ul
      className="MuiList-root MuiList-padding"
    >
      <li
        className="MuiListItem-root makeStyles-listItem-1 MuiListItem-gutters"
        disabled={false}
      >
        <div
          className="MuiListItemAvatar-root"
        >
          <span
            aria-hidden={true}
            className="material-icons MuiIcon-root"
          >
            extension
          </span>
        </div>
        <div
          className="MuiListItemText-root MuiListItemText-multiline"
        >
          <span
            className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            <a
              href="/strategies/view/Another"
              onClick={[Function]}
            >
              <strong>
                Another
              </strong>
            </a>
          </span>
          <p
            className="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
          >
            another's description
          </p>
        </div>
        <div
          aria-describedby={null}
          className=""
          onBlur={[Function]}
          onFocus={[Function]}
          onMouseLeave={[Function]}
          onMouseOver={[Function]}
          onTouchEnd={[Function]}
          onTouchStart={[Function]}
          title="Deprecate activation strategy"
        >
          <button
            className="MuiButtonBase-root MuiIconButton-root"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiIconButton-label"
            >
              <span
                aria-hidden={true}
                className="material-icons MuiIcon-root"
              >
                visibility_off
              </span>
            </span>
            <span
              className="MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          aria-describedby={null}
          className=""
          onBlur={[Function]}
          onFocus={[Function]}
          onMouseLeave={[Function]}
          onMouseOver={[Function]}
          onTouchEnd={[Function]}
          onTouchStart={[Function]}
          title="You cannot delete a built-in strategy"
        >
          <button
            className="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
            disabled={true}
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={-1}
            type="button"
          >
            <span
              className="MuiIconButton-label"
            >
              <span
                aria-hidden={true}
                className="material-icons MuiIcon-root"
              >
                delete
              </span>
            </span>
          </button>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
